src/nupic/types/Sdr.hpp

```diff
@@ -582,32 +584,55 @@ class SparseDistributedRepresentation : public Serializable
         }
         return stream << std::endl;
     }
+    friend std::istream& operator>> (std::istream& inStream, SparseDistributedRepresentation &sdr)
+    {
+        auto readVector = [&inStream] (std::vector<UInt> &vec) { 
+            inStream >> std::skipws;  // skip leading whitespace
+            vec.clear();
+            std::string val;
+            int c = 0;
+            while(c != '\n' && !inStream.eofbit) {
+              int c = inStream.get();
+              if (isdigit(c)) val.append(1, c);
+              else if (c == ',') {
+                vec.push_back( atoi(val.c_str()) );
+                val.clear();
+              }
+              else if (c != ' ') break;
+            }
+            if (!val.empty())
+              vec.push_back( atoi(val.c_str()) );
+        };
+
+        // Read the starting marker and version.
+        std::string tag;
+        inStream >> tag;
+        NTA_CHECK( tag == "SDR(" ) << "stream does not contain an SDR from <<.";
+
+        // Read the dimensions.
+        readVector( sdr.dimensions_ );
+
+        // Initialize the SDR.
+        // Calculate the SDR's size.
+        sdr.size_ = 1;
+        for(UInt dim : sdr.dimensions_)
+            sdr.size_ *= dim;
+        // Initialize sparse tuple.
+        sdr.coordinates_.assign( sdr.dimensions_.size(), {} );
```

Comment:
Nitpick: call the initialize method?

Reply:
> Nitpick: call the initialize method?

Hmmm, you mean reshape( ) maybe?

This is in the `operator>>` function which parses the output generated by `operator<<`.
There is a minor problem...I have one a lambda function that parses the stream into a `vector<UInt>`.  That works for the dimensions.  However, for the sparse data, the type is `vector<ElemSparse>` (which happens to resolve to UInt).  If I used this same lambda function for both it will compile but if ElemSparse should change then we are screwed.  We don't have templated lambda functions until C++20 so I am forced to make the lambda function into a real private function.  
Do you have any other ideas?

Reply:
`ElemSparse` is going to be a UInt32 until someone wants to make an HTM with more than 4 billion cells.  I don't think this is going to be a concern for a while...

Reply:
On second thought, I think what you have here now is fine as is.

Currently `operator>>`  constructs the SDR (assigns new dimensions).
If `operator>>` calls resize then the SDR must already be initialized with the correct size (or it will throw exception).

IMO the best would be for `operator>>` to check if the SDR is initialized (has dimensions).  If the SDR is initialized then it should check dimensions or reshape it.  If the SDR is not initialized then it should just assign the dimensions.  This how SDR assigment `operator=` currently works.